# Práctica 1 - María Godín y Alan Glotzer

## Apartado 1 - Resultados

### Tic-Tac-Toe.data

In [36]:
import EstrategiaParticionado
import Clasificador
import numpy as np
from Datos import Datos

dataset=Datos('tic-tac-toe.data')

estrategia=EstrategiaParticionado.ValidacionSimple()
clasificador=Clasificador.ClasificadorNaiveBayes()

errores=clasificador.validacion(estrategia,dataset,clasificador)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion simple sin Laplace TicTacToe -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

errores=clasificador.validacion(estrategia,dataset,clasificador, Laplace=1)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion simple con Laplace TicTacToe -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

estrategia=EstrategiaParticionado.ValidacionCruzada(4)
clasificador=Clasificador.ClasificadorNaiveBayes()

errores=clasificador.validacion(estrategia,dataset,clasificador)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion cruzada (k=4) sin Laplace TicTacToe -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

errores=clasificador.validacion(estrategia,dataset,clasificador, Laplace=1)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion cruzada (k=4) con Laplace TicTacToe -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")



- Validacion simple sin Laplace TicTacToe -
Tasa de errores:  [0.33402922755741127] 
Tasa de aciertos:  [0.6659707724425887]
Media de los errores:  0.334029227557 
Desviacion tipica de los errores:  0.0 

- Validacion simple con Laplace TicTacToe -
Tasa de errores:  [0.35073068893528186] 
Tasa de aciertos:  [0.6492693110647181]
Media de los errores:  0.350730688935 
Desviacion tipica de los errores:  0.0 

- Validacion cruzada (k=4) sin Laplace TicTacToe -
Tasa de errores:  [0.3054393305439331, 0.3598326359832636, 0.35564853556485354, 0.3651452282157676] 
Tasa de aciertos:  [0.6945606694560669, 0.6401673640167365, 0.6443514644351465, 0.6348547717842323]
Media de los errores:  0.346516432577 
Desviacion tipica de los errores:  0.0239534805223 

- Validacion cruzada (k=4) con Laplace TicTacToe -
Tasa de errores:  [0.35564853556485354, 0.36401673640167365, 0.34309623430962344, 0.3236514522821577] 
Tasa de aciertos:  [0.6443514644351465, 0.6359832635983264, 0.6569037656903766, 0.6763485477

### German.data

In [37]:
dataset=Datos('german.data')

estrategia=EstrategiaParticionado.ValidacionSimple()
clasificador=Clasificador.ClasificadorNaiveBayes()

errores=clasificador.validacion(estrategia,dataset,clasificador)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion simple sin Laplace German -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

errores=clasificador.validacion(estrategia,dataset,clasificador, Laplace=1)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion simple con Laplace German -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

estrategia=EstrategiaParticionado.ValidacionCruzada(4)
clasificador=Clasificador.ClasificadorNaiveBayes()

errores=clasificador.validacion(estrategia,dataset,clasificador)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion cruzada (k=4) sin Laplace German -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

errores=clasificador.validacion(estrategia,dataset,clasificador, Laplace=1)
aciertos = [1-x for x in errores]
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
media = np.mean(errores)

print("- Validacion cruzada (k=4) con Laplace German -\nTasa de errores: ", errores, "\nTasa de aciertos: ", aciertos)
print("Media de los errores: ", media, "\nDesviacion tipica de los errores: ", d_tipica, "\n")

- Validacion simple sin Laplace German -
Tasa de errores:  [0.306] 
Tasa de aciertos:  [0.694]
Media de los errores:  0.306 
Desviacion tipica de los errores:  0.0 

- Validacion simple con Laplace German -
Tasa de errores:  [0.326] 
Tasa de aciertos:  [0.6739999999999999]
Media de los errores:  0.326 
Desviacion tipica de los errores:  0.0 

- Validacion cruzada (k=4) sin Laplace German -
Tasa de errores:  [0.304, 0.272, 0.316, 0.3] 
Tasa de aciertos:  [0.696, 0.728, 0.6839999999999999, 0.7]
Media de los errores:  0.298 
Desviacion tipica de los errores:  0.0161245154966 

- Validacion cruzada (k=4) con Laplace German -
Tasa de errores:  [0.296, 0.292, 0.276, 0.324] 
Tasa de aciertos:  [0.704, 0.708, 0.724, 0.6759999999999999]
Media de los errores:  0.297 
Desviacion tipica de los errores:  0.0172916164658 



## Apartado 2 - Análisis

A la hora de estudiar los resultados podemos entrar en detalle en algunos aspectos de los mismos. Por un lado, al utilizar validación cruzada se obtienen valores más precisos que frente a validación simple, ya que trabaja con más resultados y obtiene una media de los mismos. Si ejecutamos el programa con el fichero TicTacToe.data, podemos obtener un resultado similar a un 34.86% de error con validación simple y un 34.65% con validación cruzada. Esta pequeña variación es debido a que la desviación presente en la desviación cruzada es mínima.

En lo relativo a Laplace, tendría un efecto notorio si hubiera abundantes datos con valor 0 en nuestras tablas pero, como no es el caso, no se aprecia una mejora muy notoria. 

Por otro lado, en lo relativo al análisis del fichero German, la media de los errores resultantes puede apreciarse que es menor que respecto a TicTacToe.data. Además, en los resultados obtenidos podemos ver un panorama similar en ambos ficheros, ya que con los datos de German.data tampoco se aprecia una desviación notoria de los errores en el caso de la validación cruzada, y tampoco cuenta con abuntantes datos con valor 0 para que Laplace signifique un cambio muy apreciable en los resultados.

## Apartado 3 - Scikit-Learn

### TicTacToe.data

In [40]:
from Datos import Datos
from sklearn import preprocessing 
from sklearn.model_selection import train_test_split 
from sklearn import naive_bayes
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict
from ClasificadorSKLearn import ValidacionSimple
from ClasificadorSKLearn import ValidacionCruzada

prediccionesSimpleSin, scoreSimpleSin = ValidacionSimple('tic-tac-toe.data', 0)
prediccionesSimpleCon, scoreSimpleCon = ValidacionSimple('tic-tac-toe.data', 1)

erroresSimpleSin = 1 - scoreSimpleSin
erroresSimpleCon = 1 - scoreSimpleCon

d_tipicaErorresSimpleSin = 0.0 #No hay en este caso
mediaErroresSimpleSin = np.mean(erroresSimpleSin)

d_tipicaErorresSimpleCon = 0.0 #No hay en este caso
mediaErroresSimpleCon = np.mean(erroresSimpleCon)

print("- Validacion simple sin Laplace TicTacToe -")
print("Tasa de errores: ", erroresSimpleSin, "\nTasa de aciertos: ", scoreSimpleSin)
print("Media de los errores: ", mediaErroresSimpleSin, "\nDesviacion tipica de los errores: ", d_tipicaErorresSimpleSin)

print("\n- Validacion simple con Laplace TicTacToe -")
print("Tasa de errores: ", erroresSimpleCon, "\nTasa de aciertos: ", scoreSimpleCon)
print("Media de los errores: ", mediaErroresSimpleCon, "\nDesviacion tipica de los errores: ", d_tipicaErorresSimpleCon)

prediccionesCruzadaSin, scoreCruzadaSin = ValidacionCruzada('tic-tac-toe.data', 0, 5)
prediccionesCruzadaCon, scoreCruzadaCon = ValidacionCruzada('tic-tac-toe.data', 1, 5)

erroresCruzadaSin = [1-x for x in scoreCruzadaSin]
erroresCruzadaCon = [1-x for x in scoreCruzadaCon]

d_tipicaErroresCruzadaSin = np.std(np.transpose(np.array(erroresCruzadaSin)), axis=0)
mediaErroresCruzadaSin = np.mean(erroresCruzadaSin)

d_tipicaErroresCruzadaCon = np.std(np.transpose(np.array(scoreCruzadaCon)), axis=0)
mediaErroresCruzadaCon = np.mean(erroresCruzadaCon)

print("\n- Validacion Cruzada sin Laplace TicTacToe -")
print("Tasa de errores: ", erroresCruzadaSin, "\nTasa de aciertos: ", scoreCruzadaSin)
print("Media de los errores: ", mediaErroresCruzadaSin, "\nDesviacion tipica de los errores: ", d_tipicaErroresCruzadaCon)

print("\n- Validacion Cruzada con Laplace TicTacToe -")
print("Tasa de errores: ", erroresCruzadaCon, "\nTasa de aciertos: ", scoreCruzadaCon)
print("Media de los errores: ", mediaErroresCruzadaCon, "\nDesviacion tipica de los errores: ", d_tipicaErroresCruzadaCon)

- Validacion simple sin Laplace TicTacToe -
Tasa de errores:  0.270833333333 
Tasa de aciertos:  0.729166666667
Media de los errores:  0.270833333333 
Desviacion tipica de los errores:  0.0

- Validacion simple con Laplace TicTacToe -
Tasa de errores:  0.295833333333 
Tasa de aciertos:  0.704166666667
Media de los errores:  0.295833333333 
Desviacion tipica de los errores:  0.0

- Validacion Cruzada sin Laplace TicTacToe -
Tasa de errores:  [0.40932642487046633, 0.29166666666666663, 0.30890052356020947, 0.39790575916230364, 0.2879581151832461] 
Tasa de aciertos:  [ 0.59067358  0.70833333  0.69109948  0.60209424  0.71204188]
Media de los errores:  0.339151497889 
Desviacion tipica de los errores:  0.0532302107295

- Validacion Cruzada con Laplace TicTacToe -
Tasa de errores:  [0.40932642487046633, 0.29166666666666663, 0.30890052356020947, 0.39790575916230364, 0.2879581151832461] 
Tasa de aciertos:  [ 0.59067358  0.70833333  0.69109948  0.60209424  0.71204188]
Media de los errores:  0.33

[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    0.0s finished
[Parallel(n_jobs=1)]: Done   5 out of   5 | elapsed:    0.0s finished


### German.data